### PR TITLE
Fix ASCII control character processing on international keyboards

### LINKED ISSUE #2462: inconsistent Ctrl-ú (Ctrl-[) handling in different modes (Czech keyboard)
After I submitted #2461 I found out that `Ctrl-ú` (which is `Ctrl-[` equivalent on Czech keyboard) does not work in the visual selection mode. For example the command `Vjjj<Ctrl-ú>` selects four lines and remains in the visual selection mode (the terminating `<Ctrl-ú>` has no effect).

When I switch to "English US" keyboard, the corresponding `Vjjj<Ctrl-[>` selects four lines and cancels the selection.

What is strange that `<Ctrl-ú>` (when using Czech layout) works ok in the insert mode, i.e. it terminates the insert mode and returns back to the normal mode. So VsVim somehow recognizes this key correctly, but not in visual selection mode.

### PR DESCRIPTION
**Pull request by ricksladkey**
#### Issues

- Closes #2462

#### Discussion

Unfortunately, this had to change the model for how `IMode.CanProcess` works and when it returns true. The following discussion is very arcane, but I'll try to describe it in detail for future reference.

As background, if VsVim declines to handle a keyboard input at all of the various input points, the input falls back to Visual Studio which would then  assume the user was typing and insert the keyboard input into the buffer. There was even an issue (issue #1793) where non-ASCII characters got inserted into the buffer in normal mode instead of beeping.

As a result, many modes in VsVim typically report that they can process many, many more inputs than are actually valid inputs for that mode. This is because they desire to process them as an error rather than let Visual Studio handle the input erroneously.

This whole approach is predicated on the premise that a VsVim mode gets one chance to process they input. But the reality is that a mode may get **multiple** chances to process the raw keystrokes typed by the user. If we decline to handle a key combination during the `KeyDown` event, that key combination **may or may not** have an alternate interpretation that appears later during the `TextInput` event. Frankly, when we are in the `KeyDown` event, we have no idea whether a later `TextInput` event will be generated or what key input it will generate. There are control keys, dead keys, magic unicode keystroke sequences, etc. that are entirely outside of our control.

The ASCII control character `0x27 / ESC` on the Czech keyboard is an example of this kind of key. During the `KeyDown` event, we encounter a composite key that does not map to any ASCII character: `Ctrl+ú`. This is a perfectly valid key input from vim's point of view. And from VsVim's for that matter. But, if we decline to handle the `KeyDown` event, we will later receive a `TextInput` event for an entirely different character, this time an ASCII one: `0x27 / ESC`. But on the US keyboard, the same thing will happen. On `KeyDown`, VsVim will see the `Ctrl+[` composite key input, and if declined, will result in `0x27 / ESC` during the `TextInput` event.

The US keyboard doesn't have any problem here because both of the key inputs are the same from vim's point of view. But on the Czech keyboard, the only way we will ever find out that the keyboard produces ASCII `0x27 / ESC` by pressing the control key down at the same time as the `ú` key, is by declining to process it during the `KeyDown` event.

The best way for VsVim to handle this is by only handling `<C-ú>` as a composite key input when the user has a mapping for that key. If there is no such mapping, and they couldn't be trying to enter text into the buffer, then we can decline to handle the composite key input because **processing it would result in an error**. More specifically, it would beep.

This means that a mode may get several opportunities to see the same keystrokes, each converted into different key inputs. Accepting one too early prevents us from see the later ones. But if we decline to handle the first one, and **there is no second one**, Visual Studio could end up inserting characters into the buffer in a non-input mode.

As a result, we now have a new `CanProcess` model:

- A mode should return `true` for `CanProcess` only if it is either a valid key or we are certain it cannot be interpreted as a different valid key
- In non-insert modes we must handle undesired insertable text **at the very last minute**, i.e. just before we would pass it on to Visual Studio

In the current Visual Studio / VsVim architecture, the point of no return for key processing is during `StandardCommandTarget.Exec`. If we return false, Visual Studio will process the key input itself. if we don't want it to do that, that's where we need to decide if we return true.

### CODE REVIEW

File: Src/VimCore/FSharpUtil.fs [hunk 1]
```diff
@@ -559,6 +559,7 @@ module internal CharUtil =
     let IsWhiteSpace x = System.Char.IsWhiteSpace(x)
     let IsNotWhiteSpace x = not (System.Char.IsWhiteSpace(x))
     let IsControl x = System.Char.IsControl x
+    let IsAscii x = x >= 0 && x <= 127
```
Comment (jaredpar) [inline Src/VimCore/FSharpUtil.fs]: Believe the type of `x` will be `int` here not `char`.

File: Src/VimCore/Modes_Insert_InsertMode.fs [hunk 2]
```diff
@@ -511,14 +512,28 @@ type internal InsertMode
...
-                    // Certain keys such as Delete, Esc, etc ... have the same behavior when invoked
-                    // with or without any modifiers.  The modifiers must be considered because they
-                    // do participate in key mapping.  Once we get here though we must discard them
+                if keyInput.KeyModifiers = VimKeyModifiers.Control then
+
+                    // A key with only the control modifier that isn't mapped
+                    // is never a direct insert. But on some international
+                    // keyboards, it might translate to an ASCII control
+                    // character. See issue #2462.
+                    match Map.tryFind keyInput _commandMap with
```
Comment (jaredpar) [inline Src/VimCore/Modes_Insert_InsertMode.fs]: When will this `match` ever succeed? It looks like a duplicate of the first `match` call in this method.

File: Src/VimCore/Modes_Normal_NormalMode.fs [hunk 3]
```diff
@@ -359,17 +356,8 @@ type internal NormalMode
...
     member x.CanProcess (keyInput: KeyInput) =
-        if _runner.IsWaitingForMoreInput then 
-            true
-        elif _runner.DoesCommandStartWith keyInput then
-            true
-        elif Option.isSome keyInput.RawChar && VimKeyModifiers.None = keyInput.KeyModifiers then
-
-            // We can process any printable character (think international input)
-            // or any character which is part of the standard Vim input set.
-            CharUtil.IsPrintable keyInput.Char || Set.contains keyInput.Char _coreCharSet
```
Comment (jaredpar) [inline Src/VimCore/Modes_Normal_NormalMode.fs]: This old code always smelled a bit to me. I'm glad you've been able to so thoroughly identify the problem here and get a better solution.

**Comment by jaredpar** [comment]
Minor feedback but overall this change looks.

Thanks for the detailed PR message here. Keyboard in put is central to VsVim and it's important to understand the why of our logic here. This particular bug has really been a puzzle to me and I'm so glad you got to the bottom of it.

File: Src/VimCore/FSharpUtil.fs (hunk 1 above)
Comment (ricksladkey) [reply Src/VimCore/FSharpUtil.fs]: Ha, ha, as much as I love type inference, it does have pitfalls for the unwary!

File: Src/VimCore/Modes_Insert_InsertMode.fs (hunk 2 above)
Comment (ricksladkey) [reply Src/VimCore/Modes_Insert_InsertMode.fs]: Ah, right. I cribbed that code as a simplification of the code directly after this section, but that other code is actually checking a different key which isn't redundant.

**Pull request merged**